llama-index-integrations/vector_stores/llama-index-vector-stores-wordlift/llama_index/vector_stores/wordlift/base.py

```diff
@@ -0,0 +1,160 @@
+import asyncio
+import logging
+import traceback
+from typing import Any, List, Dict
+
+from llama_index.core.schema import BaseNode, MetadataMode, TextNode
+from llama_index.core.vector_stores import VectorStoreQuery, VectorStoreQueryResult
+from llama_index.core.vector_stores.types import (
+    VectorStore,
+)
+
+from manager_client import NodeRequest, VectorSearchQueryRequest
+from manager_client.exceptions import ServiceException
+from utils import (
+    VectorSearchService,
+    WordliftVectorStoreException,
+    WordliftVectorQueryServiceException,
+)
+
+log = logging.getLogger("global")
+
+
+class KeyProvider:
+    key: str
+
+    def __init__(self, key: str):
+        self.key = key
+
+    async def for_add(self, nodes: List[BaseNode]) -> str:
+        return self.key
+
+    async def for_delete(self, ref_doc_id: str) -> str:
+        return self.key
+
+    async def for_query(self, query: VectorStoreQuery) -> str:
+        return self.key
+
+
+class WordliftVectorStore(VectorStore):
+    stores_text = True
+
+    vector_search_service: VectorSearchService
+
+    @staticmethod
+    def create(key: str):
+        return WordliftVectorStore(KeyProvider(key), VectorSearchService())
+
+    def __init__(
+        self,
+        key_provider: KeyProvider,
+        vector_search_service: VectorSearchService,
+    ):
+        super(WordliftVectorStore, self).__init__(use_async=True)
+
+        self.vector_search_service = vector_search_service
+        self.key_provider = key_provider
+
+    def add(self, nodes: List[BaseNode], **add_kwargs: Any) -> List[str]:
+        log.debug("Add node(s)\n")
+        loop = asyncio.new_event_loop()
+        asyncio.set_event_loop(loop)
+        task = loop.create_task(self.async_add(nodes, **add_kwargs))
+        add = loop.run_until_complete(task)
+        loop.close()
+        return add
+
+    async def async_add(
+        self,
+        nodes: List[BaseNode],
+        **kwargs: Any,
+    ) -> List[str]:
+        # Empty nodes, return empty list
+        if not nodes:
+            return []
+
+        log.debug("{0} node(s) received\n".format(len(nodes)))
+
+        # Get the key to use for the operation.
+        key = await self.key_provider.for_add(nodes)
+
+        requests = []
+        for node in nodes:
+            node_dict = node.dict()
+            metadata: Dict[str, Any] = node_dict.get("metadata", {})
+            entity_id = metadata.get("entity_id", None)
```

Comment:
is an entity id required? What happens if it's not provided?

(I really should go read more about wordlift haha)

Reply:
Wordlift Knowledge Graphs are built on the principles of fully Linked Data, where each entity is assigned a permanent dereferentiable URI.  When adding nodes to an existing Knowledge Graph, it's essential to include an "entity_id" in the metadata of each loaded document. In the future we may provide an endpoint that generates the ID

Reply:
Is there any reason we can't use `node.id_` ?

Reply:
(sorry for the naive question, this is just a bit of a UX impact)

Reply:
@logan-markewich entity IDs in WordLift are URIs following the [5 Stars Linked Data principles](https://5stardata.info/en/).

The `node.id_` doesn't (currently) satisfy this requirement, so that in this first release we require an entity ID (in the form of a URI) to be provided by the client.

We're planning to release some platform updates in the near future (end of Q3) whereas the platform can automatically generate these URIs based on different strategies (e.g. GS1 Digital Link for Products with GTIN) and we may well add a strategy that will support creating an entity URI given a `node.id_`.

Unless this is blocking, we'd rather prefer to move forward as is and then plan for future updates.